crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2646,36 +2647,125 @@ impl<'db> Type<'db> {
+                            KnownClass::Str => arguments
+                                .first_argument()
+                                .map(|arg| arg.str(db))
+                                .unwrap_or_else(|| Type::string_literal(db, "")),
+
+                            KnownClass::Type => arguments
+                                .exactly_one_argument()
+                                .map(|arg| arg.to_meta_type(db))
+                                .unwrap_or_else(|| KnownClass::Type.to_instance(db)),
+
+                            // TODO: remove a dedicated branch when `__new__` handling is added.
+                            // at that point the regular logic should be able to check signatures
+                            // of other KnownClass constructors against typeshed.
+                            _ => Type::Instance(InstanceType { class }),
+                        },
+                    )))
+                } else {
+                    // User defined class
+                    let instance_ty = Type::Instance(InstanceType { class });
+
+                    // As of now we do not model custom `__call__` on meta-classes, so the code below
+                    // only deals with interplay between `__new__` and `__init__` methods.
+                    // The logic is roughly as follows:
+                    // 1. If `__new__` is defined anywhere in the MRO (except for `object`, since it is always
+                    //    present), we call it, analyze outcome and stop. There is no need to try binding
+                    //    `__init__` in this case, since any signature mismatches should be handled at definition
+                    //    site of the class.
+                    // 2. If `__new__` is not found, we try to call `__init__`. Here, we allow it to fallback all
+                    //    the way to `object` (single argument call). This matches runtime behavior,
+                    //    where `object.__init__` would only allow >=1 arguments if `__new__` is explicitly defined.
+                    //
+                    // Note that we currently ignore `__new__` return type, since we do not support `Self`
+                    // and most builtin classes use it as return type annotation. We always return the instance type.
+
+                    // First check if there is a `__new__` method anywhere in MRO except for `object`.
+                    let check_call_outcome =
+                        if class.class_member(db, "__new__", false).symbol.is_unbound() {
+                            // No explicit `__new__` method found, try calling `__init__` on instance.
+                            instance_ty.try_call_dunder(db, "__init__", arguments)
+                        } else {
+                            // Explicit `__new__` method found, call it and return the result.
+                            self.try_call_dunder(db, "__new__", arguments)
+                        };
```

Comment:
Don't we need to call `__new__` with `&CallArguments::positional([self])`? I would expect that we can always call `__new__` using `self.try_call_dunder(db, "__new__", &CallArguments::positional([self])`, but then we probably need some special casing for `object.__new__`, because we don't understand `Self` yet.